HW1/P4/P4_bfs.py

```diff
@@ -0,0 +1,31 @@
+import time
+
+num_of_partitions = 2
+
+def copartitioned(RDD1, RDD2):
+    "check if two RDDs are copartitioned"
+    return RDD1.partitioner == RDD2.partitioner
+
+def SSBFS( source, graph, accum ): 
+    graph = graph.map( lambda ( x, y ): ( x, ( -1, y ) ) if x != source else ( x, ( 0, y ) ), preservesPartitioning = True )
+    graph.cache()
+    i = 0
+    def adder( (x, y) ):
+        accum.add(1)
+        return (x, y[1][0])
+    while( True ):
+        source = graph.filter( lambda (x,y): y[0] == i ).mapValues( lambda y: y[1] ).flatMapValues( lambda y: y )
+        assert copartitioned( graph, source )
+        accum_before = accum.value
+        not_visited = graph.filter( lambda (x,y): y[0] == -1 ).partitionBy(num_of_partitions)
+        source = source.map( lambda (x,y) : (y, 1) ).distinct().join( not_visited ).map( lambda (x,y): adder( ( x, y ) ) ).partitionBy(num_of_partitions)
```

Comment:
This line of code took me several minutes to understand. It may have been helpful to break it up into multiple lines (one of which is labeled neighbors or next_nodes or somesuch) and/or add comments about non-obvious things like using 1 as a dummy value and using adder to get rid of the dummy value.